tacc.1
.TH TACC 1
.SH NAME
tacc \- simple TACACS+ client and login program
.SH SYNOPSIS
.B tacc
[-TRAVh] [-qwn] [-u \fIusername\fR] [-p \fIpassword\fR] [-s \fIserver\fR]
[-k \fIsecret\fR] [-c \fI'command'\fR] [--username=\fIusername\fR]
[--password=\fIpassword\fR] [--server=\fIserver\fR] [--secret=\fIsecret\fR]
[--command=\fI'command'\fR | --exec=\fI'command'\fR] [--quiet | --silent]
[--no-wtmp] [--no-encrypt]
.PP
.B tacc
\fIusername\fR
.SH DESCRIPTION
.B tacc
is simple TACACS+ client, designed mainly for two purposes:
.TP
.B command line mode
sending arbitrary TACACS+ requests from command line
.TP
.B login mode
sending authentication and authorization requests for username
supplied in command line and password read from standard input;
in this case \fBtacc\fR is working in "dumb" mode and compiled-in
defaults are used
.PP
\fBtacc\fR will run in either command line mode or login mode depending
on parameters passed from command line. When there's only one non-option
(i.e. not starting with "\-") parameter, \fBtacc\fR will run in login mode.
Otherwise, it will parse other parameters described below. Any data not
explicitly set from command line will be set to default, compiled in
value.
.SH "COMMAND LINE OPTIONS"
Options specifying action to be performed. At least one is required.
.TP
.I "\-T, \-\-authenticate"
perform authentication of username and password supplied with
\-u and \-p options
.TP
.I "\-R, \-\-authorize"
request authorization for particular service (in current version
this is limited to service \fBppp\fR and protocol \fBip\fR)
.TP
.I "\-A, \-\-account"
send accounting requests on session beginning and end
.IP "Data\ options"
.TP
.I "\-h, \-\-help"
display brief listing of options and exit
.TP
.I "\-V, \-\-version"
display program name, program version and exit

.PP
Options for passing various data to the program. Only \fB\-u\fR is required.
If any of other options is not specified, the compiled-in defaul is used.

.TP
.I "\-u, \-\-username"
username for authentication, authorization and accounting
.TP
.I "\-p, \-\-password"
password for authentication; if not specified from command line,
prompt is displayed and password is read from standard input
.TP
.I "\-s, \-\-server"
TACACS+ server address, either IP address or host name; this option
can be given up to two times, specifying primary and secondary servers
.TP
.I "\-k, \-\-secret"
shared secret used to encrypt and decrypt packets exchanged with
TACACS+ server
.TP
.I "\-c, \-\-command, \-\-exec"
command to execute after successful performing all specified actions;
this usually starts PPP or SLIP driver

.PP
Options modifying program's behaviour.

.TP
.I "\-q, \-\-quiet, \-\-silent"
supresses displaying informational and error messages to standard
output, but still report errors to syslog(3)
.TP
.I "\-w, \-\-no\-wtmp"
supresses writing accounting records to local wtmp(5) files
.TP
.I "\-n, \-\-no\-encrypt"
disables encryption of packets sent to TACACS+ server

.SH "LOGIN MODE"
In this mode \fBtacc\fR will accept only one parameter, the \fIusername\fR,
as passed from \fIgetty\fR style programs. It is intended for use as
replacement for \fIlogin\fR program in \fIgetty\fR configuration file.
In login mode \fBtacc\fR will use compiled in defaults for server address,
secret key and command to run after successful authentication and
authorization. It will also prompt for password and read it from standard
input, like the \fIlogin\fR program.

.SH LIMITATIONS
In current version there is no other way to change login mode parameters
than changing them in the source file and recompiling \fBtacc\fR. It also
does not implement the full TACACS+ specification, only the subset of
TACACS+ required to perform PAP authentication, IP authorization
and start/stop accounting.
.PP
There are also some limitations in information sent in accounting packets,
caused by the fact that \fBtacc\fR doesn't perform PPP connection itself,
but only spawns \fIpppd\fR to handle the PPP protocol. Because of this
\fBtacc\fR has no idea on how many data was exchanged with the peer.

.SH "EXIT CODE"
\fBtacc\fR returns the following codes at exit:
.TP
.B 0
to indicate success
.TP
.B 1
to indicate authentication failure or remote server error
.TP
.B 2
to indicate local error

.SH "SEE ALSO"
.TP
.I ftp://ftp-eng.cisco.com/tacplus/
CISCO archive with latest versions of TACACS+ specification, API and
server source code
.TP
.I TAC_PLUS Developer's Kit
distributed with TACACS+ server source code
.TP
.I http://ceti.com.pl/~kravietz/progs/tacacs.html
more information on \fBtacc\fR and TACACS+ support for \fIpppd\fR

.SH AUTHOR
Pawel Krawczyk <kravietz@ceti.com.pl>


ChangeLog
1.4.0-1
Dave Olson, June 2016

Changes to support local mapping, so that TACACS users do not need
entries in /etc/passwd to supply home directory, uid, and gid information.

This was done by using a new mapping library libtacplus_map.  See that
package for details.

Also see the comments about immutable loginuid in Pam.d.common-example

libtac is converted to a shared library, so it can be used by other programs,
and only functions and variables starting with tac_* are exported in the
shared library.  Some functions were renamed to make this possible.

A separate package libnss_tacplus uses the mapping library to do lookups by
both name and uid.  uid lookups are only possible while a tacacs user is
logged in.

If multiple tacacs users at the same privilege level are logged in, the
current behavior is that is that if a call is done from within the login
session, the correct (login) name will be returned.   If from outside the
session (audit uid and/or session don't match in the mapping file), the name
from first map entry is used, much like normal systems where multiple users
have the same UID.

Added the runtime config capability to include another file, so that
the tacacs servers are only listed in a single place.  Ship using
/etc/tacplus_servers as an include file, and use it in the pam sample config
Because that's common, allow debug=NUMBER for pam_tacplus, as well as plain
"debug".

Renamed external libtac functions to all have a tac_ prefix, to avoid
name collision with other programs (the x*alloc family was an issue, in
particular).   This is an API change, but since library just got bumped from
1.0 to 2.0, left it at 2.0

Enabled -Werror to catch errors early (and fixed a few related items).

1.4.0
* Use openssl by default for crypto

1.3.9
* Close file descriptor leak
* Add client_connect_source_address

1.3.8
* A lot of cleanups and improvements by Walter de Jong <walter@heiho.net>
* Fixed build instruction in spec file and INSTALL
* Active_server can not be a pointer, data lost after authentication.
* Added port option per server, thanks to Luc Ducazu <lducazu@gmail.com>
* Fixed missing FIONREAD for solaris
* Rearranged header file include for libtac.h, fixes AIX compile problems
* Renamed rem_addr, rem_addr_len to r_addr and r_addr_len

1.3.7
* Tac_encryption fully handled by libtac no need to enable it manually
* Fixed connection handling in _pam_account,
  thanks to James Allwright <jamesallwright@yahoo.co.uk>
* Handle attributes which contains no value,
  thanks to James Allwright <jamesallwright@yahoo.co.uk>
* Global variables tac_login and tac_secret not static anymore,
  pointed out by James Allwright <jamesallwright@yahoo.co.uk>
* version.c: libtac version 1.8.1
* pam_tacplus.c: moved debug message after active_server validation, avoiding
  null pointer exception
* attrib.c: explicity setting *attr to NULL after free(),
  thanks to Anthony Low <anthonyl@xkl.com>

1.3.6
* Added libpam-runtime support for debian
* Added use_first_pass and try_first_pass option, thanks to Luc Ducazu <lducazu@gmail.com>
* Changed e-mail adres to jeroen@jeroennijhof.nl
* Improved accounting, added cmd attribute for command logging
* Added tac_acct_flag2str()
* Renamed tac_account_read, tac_account_send to tac_acct_read and tac_acct_send
* pam_tacplus.spec.in: fixed static library path and pam_tacplus.so location
* Debian packaging improvements

1.3.5
* This version will be dedicated to Darren Besler, thank you for your major
  contribution!

* libtac version is now 1.7.1
* magic.c: magic_inited is only used for linux
* Finally got rid of all goto illness!
* Changed tabsize to 4
* Fixed missing xalloc.h in authen_s.c
* Get PAM_RHOST from PAM stack and use it as rem_addr
* Added _pam_get_rhost() and _pam_get_user()

* The following is done by Darren Besler:
- add ability to set more elements of tacacs+ packet from parameters or globals
- cleanup messaging to be consistent with function and presentation format
- cleanup how strings are handled and returned
- acct and author read require areply.msg to be freed by caller now
- cast return values
- added port # to formatted IP address
- add timeout on read capability
- cleanup method messages are returned to caller, including adding a 0 byte
  0 byte added for safety reasons
- caller must free areply.msg now.
- add rem_addr as an argument
- include rem_addr in packet
- include ability to set priv_lvl in packet
- add ability to set authen_service from global variable aot fixed value

Bugs fixed by Darren Besler:
- cleanup various memory leaks, lost memory, and dangling pointers
- attrib.c: wasn't preserving '*' separator in attrib.c
- author_r.c:
- free attributes for replace status. Was always adding.
- uncasted char* for length was producing negative length to bcopy for arg len > 127
- possible null dereference when no separator
- cont_s.c
- was creating a new session id, should be using session id from authen start.
- magic.c
- magic was returning 0 on first call. Wasn't being initialized properly.

Other changes by Darren Besler:
* libtac/include/cdefs.h
- add #ifndef guards

* libtac/include/libtac.h
- rename #ifndef guard to match filename
- add extern "C" for C++
- alter define for TACDEBUG
- add define for TACSYSLOG
- alter macro for TACDEBUG to be able to be used at runtime via tac_debug_enable
- add declarations from tacplus.h not related to protocol
- add defines for return status codes for library functions
- add declarations for new additional global variables
tac_priv_lvl
tac_authen_method
tac_authen_service
tac_debug_enable
tac_readtimeout_enable
- revise declarations for functions to that have altered parameters lists, or return value

* libtac/include/tacplus.h
- move library specific declarations to libtac.h, leaving declarations
here to be used for protocol specific details
- add additional declarations for more complete coverage of tacacs+ protocol (v1.78)

1.3.4
* removed encrypt option just check if there is a secret (key).

tacc.c
/* tacc.c  TACACS+ PAP authentication client
 *
 * Copyright 1997-98 by Pawel Krawczyk <kravietz@ceti.com.pl>
 * Portions copyright (c) 1989 Carnegie Mellon University.
 *
 * See http://www.ceti.com.pl/~kravietz/progs/tacacs.html
 * for details.
 *
 */

#include <stdio.h>
#include <pwd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <errno.h>
#include <utmp.h>
#include <sys/file.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>
#include <time.h>
#include <getopt.h>
#include <ctype.h>
#include <openssl/rand.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "tacplus.h"
#include "libtac.h"

/* Prompt displayed when asking for password */
#define PASSWORD_PROMPT "Password: "

/* if defined, given command will be run after
 * successful authentication and proper wtmp
 * entries will be made
 */
#define DEFAULT_COMMAND "/usr/sbin/pppd -detach"

/* message that will be displayed to user
 * before starting COMMAND
 */
#define COMMAND_MESSAGE "Starting PPP\n"

/* timeout for reading password from user (seconds) */
#define GETPASS_TIMEOUT 60

/* end of CONFIGURABLE PARAMETERS */

/* prototypes */
void sighandler(int sig);
void showusage(char *argv0);
unsigned long getservername(char *serv);
void showusage(char *progname);
void showversion(char *progname);
void authenticate(struct addrinfo *tac_server, const char *tac_secret,
		const char *user, char *pass, char *tty, char *remote_addr);
void timeout_handler(int signum);

#define	EXIT_OK		0
#define	EXIT_FAIL	1	/* AAA failure (or server error) */
#define	EXIT_ERR	2	/* local error */

#define USE_SYSTEM	1

/* globals */
int tac_encryption = 1;
typedef unsigned char flag;
flag quiet = 0;
char *user = NULL; /* global, because of signal handler */
char *iface = NULL; /* -I interface or VRF to use for connection */

/* command line options */
static struct option long_options[] = {
/* operation */
    { "authenticate", no_argument, NULL, 'T' },
    { "authorize", no_argument, NULL, 'R' },
    { "account", no_argument, NULL, 'A' },
    { "version", no_argument, NULL, 'V' },
    { "help", no_argument, NULL, 'h' },

/* data */
    { "username", required_argument, NULL, 'u' },
    { "remote", required_argument, NULL, 'r' },
    { "password", required_argument, NULL, 'p' },
    { "server", required_argument, NULL, 's' },
    { "secret", required_argument, NULL, 'k' },
    { "command", required_argument, NULL, 'c' },
    { "exec", required_argument, NULL, 'c' },
    { "service", required_argument, NULL, 'S' },
    { "protocol", required_argument, NULL, 'P' },
    { "remote", required_argument, NULL, 'r' },
    { "interface", required_argument, NULL, 'I' },

/* modifiers */
    { "quiet", no_argument, NULL, 'q' },
    { "silent", no_argument, NULL, 'q' },
    { "no-wtmp", no_argument, NULL, 'w' },
    { "no-encrypt", no_argument, NULL, 'n' },
    { 0, 0, 0, 0 } };

/* command line letters */
char *opt_string = "TRAVIhu:p:s:k:c:qr:wnS:P:";

int main(int argc, char **argv) {
	char *pass = NULL;
	char *tty;
	char *command = NULL;
	char *remote_addr = NULL;
    char *service = NULL;
    char *protocol = NULL;
	struct addrinfo *tac_server;
	char *tac_server_name = NULL;
	char *tac_secret = NULL;
	int tac_fd;
	short int task_id = 0;
	char buf[40];
	int ret;
#ifndef USE_SYSTEM
	pid_t pid;
#endif
	char *msg;
	struct areply arep;

	/* options */
	flag log_wtmp = 1;
	flag do_author = 0;
	flag do_authen = 0;
	flag do_account = 0;
	flag login_mode = 0;

	/* check argc */
	if (argc < 2) {
		showusage(argv[0]);
		exit(EXIT_ERR);
	}

	/* check for login mode */
	if (argc == 2 && isalpha(*argv[1])) {
		user = argv[1];
		do_author = do_authen = do_account = 1;
		command = DEFAULT_COMMAND;
		login_mode = 1;
	} else {
		int c;
		int opt_index;

		while ((c = getopt_long(argc, argv, opt_string, long_options,
				&opt_index)) != EOF) {
			switch (c) {
			case 'T':
				do_authen = 1;
				break;
			case 'R':
				do_author = 1;
				break;
			case 'A':
				do_account = 1;
				break;
			case 'V':
				showversion(argv[0]);
			case 'h':
				showusage(argv[0]);
			case 'I':
				iface = optarg;
				break;
			case 'u':
				user = optarg;
				break;
			case 'r':
				remote_addr = optarg;
				break;
			case 'p':
				pass = optarg;
				break;
			case 's':
				tac_server_name = optarg;
				break;
			case 'k':
				tac_secret = optarg;
				break;
			case 'c':
				command = optarg;
				break;
			case 'S':
				service = optarg;
				break;
			case 'P':
				protocol = optarg;
				break;
			case 'q':
				quiet = 1;
				break;
			case 'w':
				log_wtmp = 0;
				break;
			case 'n':
				tac_encryption = 0;
				break;
			}
		}
	}

	/* check available information and set to defaults if needed */
	if (do_authen + do_author + do_account == 0) {
		printf("error: one of -TRAVh options is required\n");
		exit(EXIT_ERR);
	}

	if (user == NULL) {
		printf("error: username is required.\n");
		exit(EXIT_ERR);
	}

	if (remote_addr == NULL) {
		printf("error: remote address is required.\n");
		exit(EXIT_ERR);
	}

	if (service == NULL) {
		printf("error: service is required.\n");
		exit(EXIT_ERR);
	}

	if (protocol == NULL) {
		printf("error: protocol is required.\n");
		exit(EXIT_ERR);
	}

	if (tac_server_name == NULL) {
		printf("error: server name is required.\n");
		exit(EXIT_ERR);
	}

	struct addrinfo hints;
	memset(&hints, 0, sizeof hints);
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	ret = getaddrinfo(tac_server_name, "tacacs", &hints, &tac_server);
	if (ret != 0) {
		printf("error: resolving name %s: %s", tac_server_name,
				gai_strerror(ret));
		exit(EXIT_ERR);
	}

	if (tac_secret == NULL) {
		printf("error: server secret is required.\n");
		exit(EXIT_ERR);
	}

	if (pass == NULL) {
		signal(SIGALRM, timeout_handler);
		alarm(GETPASS_TIMEOUT);
		pass = getpass(PASSWORD_PROMPT);
		alarm(0);
		signal(SIGALRM, SIG_DFL);
		if (!strlen(pass))
			exit(EXIT_ERR);
	}

	tty = ttyname(0);
	if (strncmp(tty, "/dev/", 5) == 0)
		tty += 5;

	/* open syslog before any TACACS+ calls */
	openlog("tacc", LOG_CONS | LOG_PID, LOG_AUTHPRIV);

	if (do_authen)
		authenticate(tac_server, tac_secret, user, pass, tty, remote_addr);

	if (do_author) {
		/* authorize user */
		struct tac_attrib *attr = NULL;
		tac_add_attrib(&attr, "service", service);
		tac_add_attrib(&attr, "protocol", protocol);

		tac_fd = tac_connect_single(tac_server, tac_secret, NULL, iface);
		if (tac_fd < 0) {
			if (!quiet)
				printf("Error connecting to TACACS+ server: %m\n");
			exit(EXIT_ERR);
		}

		tac_author_send(tac_fd, user, tty, remote_addr, attr);

		tac_author_read(tac_fd, &arep);
		if (arep.status != AUTHOR_STATUS_PASS_ADD
				&& arep.status != AUTHOR_STATUS_PASS_REPL) {
			if (!quiet)
				printf("Authorization FAILED: %s\n", arep.msg);
			exit(EXIT_FAIL);
		} else {
			if (!quiet)
				printf("Authorization OK: %s\n", arep.msg);
		}

		tac_free_attrib(&attr);
	}

	/* we no longer need the password in our address space */
	bzero(pass, strlen(pass));
	pass = NULL;

	if (do_account) {
		/* start accounting */
		struct tac_attrib *attr = NULL;
		sprintf(buf, "%lu", time(0));
		tac_add_attrib(&attr, "start_time", buf);
#if defined(HAVE_OPENSSL_RAND_H) && defined(HAVE_LIBCRYPTO)
		RAND_pseudo_bytes((unsigned char *) &task_id, sizeof(task_id));
#else
        task_id = tac_magic();
#endif
		sprintf(buf, "%hu", task_id);
		tac_add_attrib(&attr, "task_id", buf);
		tac_add_attrib(&attr, "service", service);
		tac_add_attrib(&attr, "protocol", protocol);

		tac_fd = tac_connect_single(tac_server, tac_secret, NULL, iface);
		if (tac_fd < 0) {
			if (!quiet)
				printf("Error connecting to TACACS+ server: %m\n");
			exit(EXIT_ERR);
		}

		tac_acct_send(tac_fd, TAC_PLUS_ACCT_FLAG_START, user, tty, remote_addr, attr);

		ret = tac_acct_read(tac_fd, &arep);
		if(ret == 0) {
			if (!quiet)
				printf("Accounting: START failed: %s\n", arep.msg);
			syslog(LOG_INFO, "TACACS+ accounting start failed: %s", arep.msg);
		} else if (!login_mode && !quiet)
			printf("Accounting: START OK\n");

		close(tac_fd);

		tac_free_attrib(&attr);

	}

	/* log in local utmp */
#ifdef HAVE_LOGWTMP
	if (log_wtmp)
	logwtmp(tty, user, "dialup");
#endif

	if (command != NULL) {
		int ret;

		syslog(LOG_DEBUG, "starting %s for %s", command, user);

		signal(SIGHUP, SIG_IGN);
		signal(SIGTERM, SIG_IGN);
		signal(SIGINT, SIG_IGN);
		signal(SIGCHLD, SIG_IGN);

#ifdef COMMAND_MESSAGE
		printf(COMMAND_MESSAGE);
#endif

#if USE_SYSTEM
		ret = system(command);
		if (ret < 0)
			syslog(LOG_WARNING, "command failed: %m");
		else
			syslog(LOG_NOTICE, "command exit code %u", ret);
#else
		pid=fork();

		if(pid == 0) {
			/* child */

			execl(DEFAULT_COMMAND, DEFAULT_COMMAND, ARGS, NULL);
			syslog(LOG_ERR, "execl() failed: %m");
			_exit(EXIT_FAIL);
		}

		if(pid < 0) {
			/* error */
			syslog(LOG_ERR, "fork failed: %m");
			exit(EXIT_FAIL);
		}

		if(pid > 0) {
			/* parent */
			int st, r;

			r=wait(&st);
		}
#endif
	}

	if (do_account) {
		/* stop accounting */
		struct tac_attrib *attr = NULL;
		sprintf(buf, "%lu", time(0));
		tac_add_attrib(&attr, "stop_time", buf);
		sprintf(buf, "%hu", task_id);
		tac_add_attrib(&attr, "task_id", buf);

		tac_fd = tac_connect_single(tac_server, tac_secret, NULL, iface);
		if (tac_fd < 0) {
			if (!quiet)
				printf("Error connecting to TACACS+ server: %m\n");
			exit(EXIT_ERR);
		}

		tac_acct_send(tac_fd, TAC_PLUS_ACCT_FLAG_STOP, user, tty, remote_addr, attr);
		ret = tac_acct_read(tac_fd, &arep);
		if (ret == 0) {
			if (!quiet)
				printf("Accounting: STOP failed: %s", arep.msg);
			syslog(LOG_INFO, "TACACS+ accounting stop failed: %s\n", arep.msg);
		} else if (!login_mode && !quiet)
			printf("Accounting: STOP OK\n");

		close(tac_fd);

		tac_free_attrib(&attr);
	}

	/* logout from utmp */
#ifdef HAVE_LOGWTMP
	if (log_wtmp)
	logwtmp(tty, "", "");
#endif

	exit(EXIT_OK);
}

void sighandler(int sig) {
	TACDEBUG((LOG_DEBUG, "caught signal %d", sig));
}

void authenticate(struct addrinfo *tac_server, const char *tac_secret,
		const char *user, char *pass, char *tty, char *remote_addr) {
	int tac_fd;
	char *msg;
	int ret;
	struct areply arep;

	tac_fd = tac_connect_single(tac_server, tac_secret, NULL, iface);
	if (tac_fd < 0) {
		if (!quiet)
			printf("Error connecting to TACACS+ server: %m\n");
		exit(EXIT_ERR);
	}

	/* start authentication */

	if (tac_authen_send(tac_fd, user, pass, tty, remote_addr, TAC_PLUS_AUTHEN_LOGIN) < 0) {
		if (!quiet)
			printf("Error sending query to TACACS+ server\n");
		exit(EXIT_ERR);
	}

	ret = tac_authen_read(tac_fd, &arep);

	if (ret != TAC_PLUS_AUTHEN_STATUS_PASS) {
		if (!quiet)
			printf("Authentication FAILED: %s\n", arep.msg);
		syslog(LOG_ERR, "authentication failed for %s: %s", user, arep.msg);
		exit(EXIT_FAIL);
	}

	if (!quiet)
		printf("Authentication OK\n");
	syslog(LOG_INFO, "authentication OK for %s", user);

	close(tac_fd);
}

void showusage(char *progname) {
	char *a;

	a = rindex(progname, '/');
	progname = (a == NULL) ? progname : ++a;

	printf("%s -- simple TACACS+ client and login, version %u.%u.%u\n",
			progname, tac_ver_major, tac_ver_minor, tac_ver_patch);
	printf("Copyright 1997-2016 by Pawel Krawczyk <pawel.krawczyk@hush.com>\n");
	printf("Usage: %s option [option, ...]\n\n", progname);
	printf(" Action:\n");
	printf("  -T, --authenticate  perform authentication with username and password\n");
	printf("  -R, --authorize     perform authorization for requested service\n");
	printf("  -A, --account       account session beginning and end\n");
	printf("  -h, --help          display this help and exit\n");
	printf("  -V, --version       display version number and exit\n\n");
	printf(" Data:\n");
	printf("  -u, --username      remote user name\n");
	printf("  -p, --password      remote user password\n");
	printf("  -s, --server        server IP address or FQDN\n");
	printf("  -r, --remote        remote client's IP address\n");
	printf("  -S, --service       requested service (e.g. ppp)\n");
	printf("  -P, --protocol      requested protocl (e.g. ip)\n");
	printf("  -k, --secret        server encryption key\n");
	printf("  -c, --command       command to execute after successful AAA\n");
	printf("       --exec         alias for --command\n\n");
	printf(" Modifiers:\n");
	printf("  -q, --quiet         don't display messages to screen (but still\n");
	printf("      --silent        report them via syslog(3))\n");
	printf("  -w, --no-wtmp       don't write records to wtmp(5)\n");
	printf("  -n, --no-encrypt    don't encrypt AAA packets sent to servers\n\n");
    printf("Example usage:\n\n");
    printf("  tacc -TRA -u test1 -p test1 -s localhost -r 1.1.1.1 -k test1 -S ppp -P ip\n");

	exit(EXIT_ERR);
}

void showversion(char *progname) {
	char *a;

	a = rindex(progname, '/');
	progname = (a == NULL) ? progname : ++a;

	printf("%s %u.%u.%u\n", progname, tac_ver_major, tac_ver_minor,
			tac_ver_patch);
	exit(EXIT_OK);
}

unsigned long getservername(char *serv) {
	struct in_addr addr;
	struct hostent *h;

	if (inet_aton(serv, &addr) == 0) {
		if ((h = gethostbyname(serv)) == NULL) {
			herror("gethostbyname");
		} else {
			bcopy(h->h_addr, (char *)&addr, sizeof(struct in_addr));
			return(addr.s_addr);
		}
	} else
	return(addr.s_addr);

	return (-1);
}

void timeout_handler(int signum) {
	syslog(LOG_ERR, "timeout reading password from user %s", user);

}
